### PR TITLE
Do not remove custom errors in IR codegen when strip revert strings is requested

### LINKED ISSUE #16465: Compiler option `revert-strings` with value `strip` has different treatments in legacy and IR pipeline
## Description
The compiler option `revert-strings` in CLI or `revertStrings` in standard JSON Input/Output when configured with the value "strip" produces different code depending on the legacy or IR pipeline.
While the IR pipeline removes strings and custom errors from the generated code for `require`, the legacy pipeline will only remove strings.
In both pipelines, the branch for the false condition of `require` generates a `revert`, so the behavior of the code is the same, but legacy pipeline will not suppress the custom error. This creates a "semantic" difference between the pipelines.

## Steps to Reproduce
Use this Solidity code

```solidity
error MyError(string emsg);
contract C {
    function f() pure external {
        require(false, MyError("abc"));
    }
}
```

compile with both pipelines and then check the generated code to verify that in legacy the custom error is present as argument to a `revert` instruction:

### Legacy pipeline
`solc test.sol --debug-info none --revert-strings strip --asm`

generated assembly:
```
  mstore(0x40, 0x80)
  callvalue
  dup1
  iszero
  tag_1
  jumpi
  revert(0x00, 0x00)
tag_1:
  pop
  dataSize(sub_0)
  dup1
  dataOffset(sub_0)
  0x00
  codecopy
  0x00
  return
stop

sub_0: assembly {
      mstore(0x40, 0x80)
      callvalue
      dup1
      iszero
      tag_1
      jumpi
      revert(0x00, 0x00)
    tag_1:
      pop
      jumpi(tag_2, lt(calldatasize, 0x04))
      shr(0xe0, calldataload(0x00))
      dup1
      0x26121ff0
      eq
      tag_3
      jumpi
    tag_2:
      revert(0x00, 0x00)
    tag_3:
      tag_4
      tag_5
      jump	// in
    tag_4:
      stop
    tag_5:
      mload(0x40)
      shl(0xe0, 0x8b3d2d43)
      dup2
      mstore
      0x20
      0x04
      dup3
      add
      mstore
      0x03
      0x24
      dup3
      add
      mstore
      shl(0xe8, 0x616263)
      0x44
      dup3
      add
      mstore
      0x64
      add
      mload(0x40)
      dup1
      swap2
      sub
      swap1
      revert

    auxdata: 0xa26469706673582212201cbe11542190a594b908ff4413cc2ee54cfe3d65d7c38cf79ccf95d10fdfc24064736f6c63430008210033
}
```

### IR pipeline
`solc test.sol --debug-info none --revert-strings strip --asm --via-ir`

generated assembly:
```
EVM assembly:
  0x80
  dup1
  0x40
  mstore
  jumpi(tag_1, callvalue)
  dataSize(sub_0)
  swap1
  dup2
  dataOffset(sub_0)
  dup3
  codecopy
  return
tag_1:
  0x00
  dup1
  revert
stop

sub_0: assembly {
      0x00
      dup1
      revert

    auxdata: 0xa2646970667358221220bd58da9e3f0aafbdbaf47ad5b1a5e02d60030b3e4811bc3e1aa83946e06b77ca64736f6c63430008210033
}

```

### PR DESCRIPTION
**Pull request by matheusaaguiar**
Fix #16465.

### CODE REVIEW

File: test/cmdlineTests/revert_strings_strip_legacy/input.sol [hunk 1]
```diff
@@ -0,0 +1,8 @@
+pragma solidity >=0.0;
+// SPDX-License-Identifier: GPL-3.0
+contract C {
+    error MyError(string errorMsg);
```
Comment (nikola-matic) [inline test/cmdlineTests/revert_strings_strip_legacy/input.sol]: I'd give this an argument or two more, e.g. `error MyError(string errorMsg, uint256 errorId);`

File: test/cmdlineTests/revert_strings_strip_legacy/input.sol [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    function f() pure external {
+        require(false, MyError("error"));
```
Comment (nikola-matic) [inline test/cmdlineTests/revert_strings_strip_legacy/input.sol]: For the callsite, declare variables, assign them values, and pass them as arguments to the error constructor call.

```solidity
string errorMsg = "error";
uint256 errorId = 111;

require(false, MyError(errorMsg, errorId));
```

I wanna see some pushes to the stack.

**Comment by cameel** [comment]
We need to also adjust documentation as a part of this issue.

And consider changing the option or value names to make it more intuitive.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -1298,15 +1298,20 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
-					solAssert(errorDefinition && errorDefinition->functionType(true));
-					utils().revertWithError(
-						errorDefinition->functionType(true)->externalSignature(),
-						errorDefinition->functionType(true)->parameterTypes(),
-						errorConstructorArgumentTypes
-					);
-					// Here, the argument is consumed, but in the other branch, it is still there.
-					m_context.adjustStackOffset(static_cast<int>(sizeOfErrorArguments));
+					if (m_context.revertStrings() == RevertStrings::Strip)
+						m_context.appendRevert();
```
Comment (nikola-matic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: My suspicion is that you'll leave the stack dirty here (i.e. when we iterate and handle the constructor call arguments (loop at 1279), we'll also push any non literal arguments to the stack, which will not be popped anywhere.

Take a look at the [revert with error(string)](https://github.com/argotorg/solidity/blob/develop/libsolidity/codegen/ExpressionCompiler.cpp#L876-L901) implementation.

**Review by nikola-matic — COMMENTED** [review]
The asm output test is definitely necessary, but please add more semantic test coverage; there's the `// revertStrings: strip` option for this.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol
Comment (rodiazet) [inline test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol:1]: Add:
```
pragma solidity >=0.0;
// SPDX-License-Identifier: GPL-3.0
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -1174,6 +1175,8 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		}
 		else
 		{
+			if (m_context.revertStrings() == RevertStrings::Strip)
```
Comment (rodiazet) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1176]: I would add here a comment explaining that revert string behaves differently for custom error and for a raw message.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -1157,10 +1157,11 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
 		solAssert(arguments.size() <= 2, "Expected no more than two parameters for require/assert");
 
-		Type const* messageArgumentType =
-			arguments.size() > 1 && m_context.revertStrings() != RevertStrings::Strip ?
-			arguments[1]->annotation().type :
-			nullptr;
+		// Type const* messageArgumentType =
+		// 	arguments.size() > 1 && m_context.revertStrings() != RevertStrings::Strip ?
+		// 	arguments[1]->annotation().type :
+		// 	nullptr;
```
Comment (rodiazet) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Remember to remove this.

File: Changelog.md [hunk 6]
```diff
@@ -5,7 +5,7 @@ Language Features:
 Compiler Features:
 
 Bugfixes:
-
+* Code Generator: Remove custom error arguments of `require` when option `revert-strings=strip` is selected in the legacy pipeline.
```
Comment (rodiazet) [inline Changelog.md]: This should be updated. It's not removing but preserving the custom error arguments.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol [hunk 7]
```diff
@@ -0,0 +1,13 @@
+contract C {
+    error MyError(uint errorCode, string errorMsg, bool flag);
+    function flag() pure public returns (bool) { return false; }
+    function f() pure external {
+        uint code = 8;
+        string memory eMsg = "error";
+        require(false, MyError(code, eMsg, flag()));
```
Comment (rodiazet) [inline test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol]: How about tests where `require`'s first arguments cannot be evaluated at compile time? Do you think it makes sense to add them?

File: test/libsolidity/semanticTests/errors/require_custom_error_constructor_side_effect_revert_strings_strip.sol [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        string memory eMsg = "error";
+        require(false, MyError(genCode(), eMsg));
+        counter = 42;
+    }
+}
+// ====
+// revertStrings: strip
+// ----
+// f() -> FAILURE, hex"0b9344e2", hex"0000000000000000000000000000000000000000000000000000000000000001", hex"0000000000000000000000000000000000000000000000000000000000000040", hex"0000000000000000000000000000000000000000000000000000000000000005", hex"6572726f72000000000000000000000000000000000000000000000000000000"
+// counter() -> 0
```
Comment (rodiazet) [inline test/libsolidity/semanticTests/errors/require_custom_error_constructor_side_effect_revert_strings_strip.sol:17]: Here and in other tests I would add an explanation that the counter returns 0 because the transaction failed and the most important check is that the error argument is properly evaluated to 1.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 9]
```diff
@@ -1158,10 +1158,7 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
 		solAssert(arguments.size() <= 2, "Expected no more than two parameters for require/assert");
 
-		Type const* messageArgumentType =
-			arguments.size() > 1 && m_context.revertStrings() != RevertStrings::Strip ?
-			arguments[1]->annotation().type :
-			nullptr;
+		Type const* messageArgumentType = arguments.size() > 1 ? arguments[1]->annotation().type : nullptr;
 
 		auto const* magicType = dynamic_cast<MagicType const*>(messageArgumentType);
 		if (magicType && magicType->kind() == MagicType::Kind::Error)
```
Comment (rodiazet) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1164]: In the new version this fragment of the code also generates code for path when the require conditions evaluates to true, but I cannot find tests for this case except `require_error_stack_check.sol`. Not sure if it tests `revertStrings = Strip`. IMO we should also add test which does not revert and check that the stack is properly cleaned and not revert is called.
BTW in previous version with Strip, custom Error was handled by the code in the `else` branch. Was it correct?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1168]: Does it make sense to add test case where there is zero custom error arguments to make sure that this corner case (zero for loop iterations) also works properly? 
`error MyError(); require(false, MyError())` and probably `error MyError(); require(true, MyError())`

Comment (rodiazet) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1]: 1. Would also add a test case, in the same contract, with stripped plain string message and un-stripped custom error message,
2. `require_error_evaluation_order_2.sol` tests proper order of require arguments evaluation. Does it also test this in case of `revertStrings: Strip`? Probably not. 
3. Similar `require_inherited_error.sol` also is tested only for non-strip version.
4. Same for `require_error_condition_evaluated_only_once.sol` it's only tested for non-strip version, and it should be tested because it old code had different control flow.

**Review by rodiazet — CHANGES_REQUESTED** [review]
LGTM in general. I found a couple of small issues mainly with test coverage. There is also one question I have regarding the way how the old version worked.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol:1]: It does not hurt, but usually we don't add that to tests... not sure if there is a real reason, but I guess we don't care.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol (hunk 7 above)
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol]: Hum, I don't think that matters or affects the codegen but I will double-check.
Can at least add that in some tests to cover for it, but I don't think a specific test is necessary.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol:1]: Ah, I realize why you suggested that. I copied these tests from the cli tests and there we actually care about that to silence warnings in the output file.
For syntax and semantic tests, that's not the case, i.e, they are filtered out.

File: test/libsolidity/semanticTests/errors/require_custom_error_constructor_side_effect_revert_strings_strip.sol (hunk 8 above)
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_custom_error_constructor_side_effect_revert_strings_strip.sol:17]: Added the explanation.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 9 above)
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1164]: > In the new version this fragment of the code also generates code for path when the require conditions evaluates to true, but I cannot find tests for this case except require_error_stack_check.sol. Not sure if it tests revertStrings = Strip.

The new version now follows the same path of previous version when revertStrings != Strip, so it should have the same result. I will add a variation of that test with `revert-strings=string` anyway. 

> BTW in previous version with Strip, custom Error was handled by the code in the else branch. Was it correct?

The handling was the same of a `require` with no message, i.e., `YulUtilFunctions::requireOrAssertFunction` will check if the message type is null and then generate code with `revert(0, 0)` for the false branch of the require condition.
The arguments of `require` are already visited and their code generated by this point (side-effects guaranteed to happen).
So I think it was correct.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1168]: Added.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 10]
```diff
@@ -1175,6 +1172,9 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		}
 		else
 		{
+			// This option only remove strings, not custom errors
```
Comment (clonker) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			// This option only removes strings, not custom errors
```

Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Fixed.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1]: Added

**Comment by matheusaaguiar** [comment]
I think I addressed all corrections and suggestions.
I added a brief explanation of the behavior of the option in the existing documented example in the Standard JSON Input Output section. I have started a separated [PR](https://github.com/argotorg/solidity/pull/16562) to refine that and have the options and their explanation better organized.

File: docs/using-the-compiler.rst [hunk 11]
```diff
@@ -378,7 +378,9 @@ Input Description
           // How to treat revert (and require) reason strings. Settings are
           // "default", "strip", "debug" and "verboseDebug".
           // "default" does not inject compiler-generated revert strings and keeps user-supplied ones.
-          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects
+          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects.
+          // NOTE: "strip" does not remove custom errors.
+          // WARNING: Before Solidity 0.8.35, "strip" in the IR pipeline used to remove custom errors.
```
Comment (clonker) [inline docs/using-the-compiler.rst]:
```suggestion
          // WARNING: Before Solidity 0.8.35, "strip" in the IR pipeline also removed custom errors.
```

matter of taste

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 12]
```diff
@@ -1158,10 +1158,7 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		solAssert(arguments.size() > 0, "Expected at least one parameter for require/assert");
 		solAssert(arguments.size() <= 2, "Expected no more than two parameters for require/assert");
 
-		Type const* messageArgumentType =
-			arguments.size() > 1 && m_context.revertStrings() != RevertStrings::Strip ?
-			arguments[1]->annotation().type :
-			nullptr;
+		Type const* messageArgumentType = arguments.size() > 1 ? arguments[1]->annotation().type : nullptr;
```
Comment (clonker) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		Type const* messageArgumentType = arguments.size() == 2 ? arguments[1]->annotation().type : nullptr;
```

I think this is slightly cleaner. You already assert that number of arguments must be either 1 or 2. But of course this departs from the original implementation a little bit more. Entirely optional change.

**Review by clonker — DISMISSED** [review]
It looks good to me and as far as I can tell, all reviewer comments have been addressed. Two tiny suggestions, feel free to ignore. I guess @nikola-matic and/or @rodiazet might want to check that for their own comments, too. Not sure if @cameel wants to have a look before this goes in.

**Comment by cameel** [comment]
> Not sure if @cameel wants to have a look before this goes in.

Yes, I was planning to take a look.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 12 above)
Comment (matheusaaguiar) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Agree. This is more precise and closely follows the asserted assumptions.

File: Changelog.md [hunk 13]
```diff
@@ -19,6 +19,7 @@ Compiler Features:
 Bugfixes:
 * Yul: Fix incorrect serialization of Yul object names containing double quotes and escape sequences, producing output that could not be parsed as valid Yul.
 * Yul EVM Code Transform: Improve stack shuffler performance by fixing a BFS deduplication issue.
+* Yul IR Code Generation: Preserve custom error argument of `require` when option `revert-strings=strip` is selected.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
* Yul IR Code Generation: Preserve custom error argument of `require` when stripping of revert strings is selected via `--revert-strings strip`.
```

File: docs/using-the-compiler.rst [hunk 14]
```diff
@@ -378,7 +378,9 @@ Input Description
           // How to treat revert (and require) reason strings. Settings are
           // "default", "strip", "debug" and "verboseDebug".
           // "default" does not inject compiler-generated revert strings and keeps user-supplied ones.
-          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects
+          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects.
+          // NOTE: "strip" does not remove custom errors.
+          // WARNING: Before Solidity 0.8.35, "strip" in the IR pipeline also removed custom errors.
```
Comment (cameel) [inline docs/using-the-compiler.rst]: That's not accurate. It would only remove them from `require()`. They were still preserved in the `revert` statement.

Comment (cameel) [reply docs/using-the-compiler.rst]: I'm not sure it's actually worth mentioning here though. It was a bug, not a design change. The entry in the changelog is probably enough.

File: test/libsolidity/semanticTests/errors/require_inherited_error_revert_strings_strip.sol [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+{
+    function f() external pure
+    {
+        require(false, CustomError(1, "two", 3));
+    }
+}
+// ====
+// revertStrings: strip
+// ----
+// f() -> FAILURE, hex"11a1077e", hex"0000000000000000000000000000000000000000000000000000000000000001", hex"0000000000000000000000000000000000000000000000000000000000000060", hex"0000000000000000000000000000000000000000000000000000000000000003", hex"0000000000000000000000000000000000000000000000000000000000000003", hex"74776f0000000000000000000000000000000000000000000000000000000000"
```
Comment (cameel) [inline test/libsolidity/semanticTests/errors/require_inherited_error_revert_strings_strip.sol]: The way isoltest formats the expectation by default sometimes makes it hard to tell at a glance if they are correct. Especially ASCII-decoding strings in your head is not fun :) Please reformat them into something more sensible:

```suggestion
// f() -> FAILURE, hex"11a1077e", 1, 0x60, 3, 3, "two"
```

In other tests as well.

File: test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    function h() internal returns (uint256) { return y; }
+    function f(bool c) public {
+        require(g(c), E(h()));
+    }
+}
+// ====
+// revertStrings: strip
+// ----
+// f(bool): false -> FAILURE, hex"002ff067", 42
+// f(bool): true ->
```
Comment (cameel) [inline test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol]: These expectations only verify the order in the revert case. Would be more interesting if the effect of `h()` was to flip the condition :)

File: test/cmdlineTests/revert_strings_strip_via_ir_require_custom_error_constructor_param_side_effect/input.sol [hunk 17]
```diff
@@ -0,0 +1,12 @@
+pragma solidity >=0.0;
+// SPDX-License-Identifier: GPL-3.0
+contract C {
```
Comment (cameel) [inline test/cmdlineTests/revert_strings_strip_via_ir_require_custom_error_constructor_param_side_effect/input.sol]: Not that it matters much, but this being unnecessarily different just distracts from the content of the test :)
```suggestion
// SPDX-License-Identifier: GPL-3.0
pragma solidity *;

contract C {
```

File: test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json
Comment (cameel) [inline test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json:1]: Our naming convention for Standard JSON test is to start the name with `standard_`.

Though I don't really think we need Standard JSON variants here that much. The issue was in the codegen, not in option processing and the output is long so it would have been just fine to only include the CLI tests.

**Review by cameel — COMMENTED** [review]
The docs need a correction - we were not always removing the custom errors. Just in `require()`.

Other than that there are a few minor things that could be cleaned up, but no real blockers.

File: test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol
Comment (cameel) [reply test/libsolidity/semanticTests/errors/require_custom_error_revert_strings_strip.sol:1]: > not sure if there is a real reason, but I guess we don't care.

We do. It's just that this is already a part of a preamble that isoltest adds automatically. No need to explicitly include it in the test.

File: docs/using-the-compiler.rst [hunk 18]
```diff
@@ -378,7 +378,9 @@ Input Description
           // How to treat revert (and require) reason strings. Settings are
           // "default", "strip", "debug" and "verboseDebug".
           // "default" does not inject compiler-generated revert strings and keeps user-supplied ones.
-          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects
+          // "strip" removes all revert strings (if possible, i.e. if literals are used) keeping side-effects.
+          // NOTE: "strip" does not remove custom errors.
```
Comment (cameel) [inline docs/using-the-compiler.rst:382]: Actually, might be good to clarify that in `--help` as well.

File: test/libsolidity/semanticTests/errors/require_inherited_error_revert_strings_strip.sol (hunk 15 above)
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_inherited_error_revert_strings_strip.sol]: Done.

File: test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol (hunk 16 above)
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol]: Done.

File: solc/CommandLineParser.cpp [hunk 19]
```diff
@@ -648,7 +648,8 @@ General Information)").c_str(),
 		(
 			g_strRevertStrings.c_str(),
 			po::value<std::string>()->value_name(util::joinHumanReadable(g_revertStringsArgs, ",")),
-			"Strip revert (and require) reason strings or add additional debugging information."
+			"Strip revert (and require) reason strings or add additional debugging information. "
+			"'strip' does not remove custom errors."
```
Comment (cameel) [inline solc/CommandLineParser.cpp]:
```suggestion
			"Note that 'strip' does not remove custom errors."
```

File: test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json
Comment (cameel) [reply test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json:1]: I think you removed too much :) It looks like all commandline tests are gone. The CLI ones were fine. I only meant the ones for Standard JSON.

File: test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol (hunk 16 above)
Comment (cameel) [reply test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol]: The way you changed it we still don't know what the evaluation order is in the non-reverting case because we can't tell what the value of `y` at the point where we evaluate `E(y)`. I was thinking about something along these lines:

```solidity
contract C {
    error E(uint256);

    uint counter = 0;

    function even() internal returns (bool) {
        return y % 2 == 0;
    }

    function inc() internal returns (uint256) {
        return ++counter;
    }

    function flip() public {
        require(even(), E(inc()));
    }
}
```

Here evaluating the second argument changes the condition in `require()`. Wrong evaluation order will be visible in both cases by changing whether we revert.

Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/errors/require_error_evaluation_order_revert_strings_strip.sol]: I see now. Thanks for the clarification.

File: test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json
Comment (matheusaaguiar) [reply test/cmdlineTests/revert_strings_strip_standard_json_io_legacy_require_custom_error/input.json:1]: Ah, true! I mistakenly removed all cli ones. Restored.

**Review by cameel — APPROVED** [review]
